File: web/lib/watchlist.ts
import { promises as fs } from "node:fs";
import path from "node:path";
import { cache } from "react";
import {
  ANALYSIS_API_CONFIGURED,
  assertAnalysisApiConfiguration,
  loadWatchlistFromApi,
} from "./api";
import { FETCH_REVALIDATE_SECONDS } from "./site";
import type { WatchlistEntry } from "./types/watchlist";

const WATCHLIST_FILE = process.env.STOCK_TICKERS_FILE
  ? path.resolve(process.env.STOCK_TICKERS_FILE)
  : path.resolve(process.cwd(), "..", "tickers.txt");
const SUPABASE_URL =
  process.env.SUPABASE_URL ?? process.env.NEXT_PUBLIC_SUPABASE_URL ?? "";
const SUPABASE_KEY =
  process.env.SUPABASE_PUBLISHABLE_KEY ??
  process.env.NEXT_PUBLIC_SUPABASE_PUBLISHABLE_KEY ??
  process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY ??
  "";

const THEME_MARKER = /^#\s*theme:\s*(.+?)\s*$/i;

/**
 * Parse tickers.txt into structured entries.
 *
 * The file stays the single source of truth shared with fetch.py; theme is
 * carried by `#theme:` comment markers, which fetch.py skips as ordinary
 * comments. Legacy `#group:` comments are ignored. `@universe` directives are
 * ignored here — expanding them needs a network fetch, and this watchlist
 * deliberately lists tickers by hand.
 */
export function parseWatchlist(text: string): WatchlistEntry[] {
  const entries: WatchlistEntry[] = [];
  const seen = new Set<string>();
  let theme: string | null = null;

  for (const raw of text.split(/\r?\n/)) {
    const line = raw.trim();
    if (!line) continue;
    if (line.startsWith("#")) {
      const t = THEME_MARKER.exec(line);
      if (t) theme = t[1];
      continue;
    }
    if (line.startsWith("@")) continue;

    const isMY = line.toUpperCase().startsWith("MY:");
    const symbol = (isMY ? line.slice(3) : line).trim().toUpperCase();
    if (!symbol || seen.has(symbol)) continue;
    seen.add(symbol);
    entries.push({ symbol, market: isMY ? "MY" : "US", theme });
  }
  return entries;
}

export const loadWatchlist = cache(async (): Promise<WatchlistEntry[]> => {
  assertAnalysisApiConfiguration();
  if (ANALYSIS_API_CONFIGURED) return loadWatchlistFromApi();
  if (SUPABASE_URL && SUPABASE_KEY) {
    const response = await fetch(
      `${SUPABASE_URL.replace(/\/$/, "")}/rest/v1/tickers?select=symbol,market,theme&enabled=eq.true&order=symbol.asc`,
      {
        headers: {
          apikey: SUPABASE_KEY,
          Authorization: `Bearer ${SUPABASE_KEY}`,
        },
        next: { revalidate: FETCH_REVALIDATE_SECONDS },
      },
    );
    if (!response.ok) {
      throw new Error(`Supabase watchlist failed (${response.status})`);
    }
    const rows = (await response.json()) as Array<{
      symbol: string;
      market: "US" | "MY";
      theme: string | null;
    }>;
    return rows.map((row) => ({
      symbol: row.symbol,
      market: row.market,
      theme: row.theme,
    }));
  }
  try {
    return parseWatchlist(await fs.readFile(WATCHLIST_FILE, "utf8"));
  } catch {
    return [];
  }
});

export const loadWatchlistMap = cache(
  async (): Promise<Record<string, WatchlistEntry>> =>
    Object.fromEntries((await loadWatchlist()).map((e) => [e.symbol, e])),
);

File: src/stock_analysis/data/watchlist.py
"""Parse hand-curated ticker and theme markers from ``tickers.txt``.

The file is shared with ``web/lib/watchlist.ts``. Both parsers keep the
``#theme:`` marker semantics aligned; ``#group:`` comments are intentionally
ignored as legacy comments.

``@universe`` directives are skipped here for the same reason the TS parser
skips them: expanding one needs a network call, and universe members are not
hand-curated watchlist entries.
"""

from __future__ import annotations

import re
from pathlib import Path

from pydantic import BaseModel

THEME_MARKER = re.compile(r"^#\s*theme:\s*(.+?)\s*$", re.IGNORECASE)


class WatchlistEntry(BaseModel):
    symbol: str
    market: str
    theme: str | None = None


def parse_watchlist(text: str) -> list[WatchlistEntry]:
    """Return hand-listed entries with their current theme context.

    Theme markers persist until the next theme marker. The symbol is the
    display symbol (``MY:1155`` -> ``1155``), which is what ``TickerInfo.symbol``
    and therefore ``tickers.symbol`` carry.
    """
    entries: list[WatchlistEntry] = []
    seen: set[str] = set()
    theme: str | None = None

    for raw in text.splitlines():
        line = raw.strip()
        if not line:
            continue
        if line.startswith("#"):
            match = THEME_MARKER.match(line)
            if match:
                theme = match.group(1)
            continue
        if line.startswith("@"):
            continue

        is_my = line.upper().startswith("MY:")
        symbol = (line[3:] if is_my else line).strip().upper()
        if not symbol or symbol in seen:
            continue
        seen.add(symbol)
        entries.append(
            WatchlistEntry(
                symbol=symbol,
                market="MY" if is_my else "US",
                theme=theme,
            )
        )
    return entries


def load_watchlist_map(path: Path | str) -> dict[str, WatchlistEntry]:
    """Return ``{symbol: entry}``, empty when the file is absent.

    A missing watchlist is not an error: universe-driven fetches
    (``@us-major``) legitimately run without a hand-curated file.
    """
    file_path = Path(path)
    if not file_path.exists():
        return {}
    return {entry.symbol: entry for entry in parse_watchlist(file_path.read_text())}
